vvau.hpp

```diff
@@ -53,6 +53,113 @@
 /**
  * \brief Vector vector activate function
  *
+ * The function performs the multiplication between a weight vector and the input activation vector,
+ * accumulating the results and then applying an activation function on the accumulated result.
+ * It is used to implement depth-wise separable convolution
+ * 
+ * \tparam Channels   Number of channels
+ * \tparam Kernel_2   Kernel * Kernel dimension (Kernel ^ 2 if square)
+ * \tparam SIMD       Number of input columns computed in parallel
+ * \tparam PE         Number of output rows computed in parallel
+ * \tparam MMV        Number of output pixels computed in parallel
+ * \tparam TSrcI      DataType of the input activation (as used in the MAC)
+ * \tparam TDstI      DataType of the output activation (as generated by the activation)
+ * \tparam TWeightI   DataType of the weights (as used in the MAC)
+ * \tparam TI         DataType of the input stream - safely deducible from the paramaters
+ * \tparam TO         DataType of the output stream - safely deducible from the paramaters
+ * \tparam TW         DataType of the weights matrix - safely deducible from the paramaters
+ * \tparam TA         DataType of the activation class (e.g. thresholds) - safely deducible from the paramaters
+ * \tparam R          Datatype for the resource used for FPGA implementation of the MAC  - safely deducible from the paramaters
+ *
+ * \param in          Input stream
+ * \param out         Output stream
+ * \param weights     Weights matrix (currently supports BinaryWeights or FixedPointWeights)
+ * \param activation  Activation class
+ * \param reps        Number of time the function has to be repeatedly executed (e.g. number of images)
+ * \param r           Resource type for the hardware implementation of the MAC block
+ */
+template<
+	unsigned Channels, unsigned Kernel_2, unsigned SIMD, unsigned PE, unsigned MMV,
+	typename TSrcI = Identity, typename TDstI = Identity, typename TWeightI = Identity,
+	typename TI, typename TO, typename TW, typename TA, typename R
+>
+void Vector_Vector_Activate_Batch(
+	hls::stream<TI> &in,
+	hls::stream<TO> &out,
+	TW  const &weights,
+	TA  const &activation,
+	int const  reps,
+	R const &r
+) {
+
+	// how many different rows each neuron will compute
+	// alternatively: number of vertical matrix chunks
+	constexpr unsigned  NF = Channels / PE;
+
+	// how many synapse groups each row is split into
+	// alternatively: number of horizontal matrix chunks
+	constexpr unsigned  SF = (Channels*Kernel_2) / Channels;
```

Comment:
What is the rationale for going forth and then back again?

Reply:
I don't think there is any good rationale, perhaps SIMD was part of the equation at some point and was partially simplifed.